proposals/text/3665-multi-task-component.md

```diff
@@ -0,0 +1,449 @@
+  a wider range of NLP tasks. This can provide users with more flexibility and control over the model,
+  and can enable them to tailor it to their specific needs and applications.
+
+  # Detailed design
+
+- The PromptNode class is the most important abstraction in Haystack's large language model (LLM) support.
+  In addition to PromptNode class, we'll also define a set of prompt templates for the most popular NLP tasks.
+
+
+- NLP prompt templates will be represented by `PromptTemplate` class.
+
+  ``` python
+      class PromptTemplate(BaseTemplate):
+
+          name: str
+          prompt_text: str
+          input_variables: List[str]
+
+  ````
+
+  PromptNode would, out-of-the-box, support 10-20 default NLP tasks defined by PromptTemplate instances. However, it would
+  allow registering additional templates with PromptNode.
+
+
+- The prompt templates for default tasks (question-answering,question-generation, summarization etc.) could be examined by the user
+  using `get_prompt_templates_names` class method of the PromptNode. For example:
+
+  ``` python
+	  from haystack.nodes.llm import PromptNode
+      PromptNode.get_prompt_templates_names()
+
+      ----------------------------------------------------------------------------
+      ['question-answering',
+       'question-generation',
+       'conditioned-question-generation',
+       'summarization',
+       'question-answering-check']
+  ```
+
+
+- PromptNode supports natural language prompting (using `prompt` method) by specifying prompt template method parameter. For example:
+
+  ``` python
+      from haystack.nodes.llm import PromptNode
+      pn = PromptNode(model_name_or_path="google/flan-t5-base")
+	  pn.prompt("question-generation", documents=["Berlin is the capital of Germany."])
+
+      ----------------------------------------------------------------------------
+      ['What is the capital of Germany?']
+  ```
+
+- PromptNode supports selecting a particular default template for a certain task (e.g. question-generation) and then subsequently
+  using the selected template until user changes the current template. For example:
+
+  ``` python
+	  qa = pn.use_prompt_template("deepset/question-generation-v2")
+      qa(documents=["Berlin is the capital of Germany."])
+
+      ----------------------------------------------------------------------------
+      ['What is the capital of Germany?']
+  ```
+
+- The addition of new prompt templates is supported by the `add_prompt_template` method. For example:
+
+  ``` python
+      from haystack.nodes.llm import PromptNode
+      PromptNode.add_prompt_template(PromptTemplate(name="sentiment-analysis",
+                              prompt_text="Please give a sentiment for this context. Answer with positive, "
+                              "negative or neutral. Context: $documents; Answer:",
+                              input_variables=["documents"]))
```

Comment:
nit: I know we listed `input_variables` today as an option, but do we really need it? What would be the value compared to "inferring" it automatically from the prompt text where we already marked $documents as the input variable

Reply:
Yes we could; it was there primarily to remove the inferring automatically code part.